clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2234,8 +2232,29 @@ static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
 
     return eraseAnonNamespace(ArgOS.str().str());
   }
-
-  return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
+  if (const auto *D = dyn_cast<CXXRecordDecl>(RD)) {
+    std::string NSStr = "";
+    const DeclContext *DC = D->getDeclContext();
+    SmallString<64> Buf;
+    llvm::raw_svector_ostream OS(Buf);
+    while (DC) {
```

Comment:
Instead of this, check out: https://clang.llvm.org/doxygen/classclang_1_1DeclContext.html#a52ebc03bfefa68fcda366dd5a7164c65

That way you can just do this in order.

Reply:
Not sure. I am assuming that you want me replace the while loop with something like this:
mallVector<DeclContext *, 8> Contexts;
    const_cast<DeclContext *>(D->getDeclContext())->collectAllContexts(Contexts);
    for (auto *DC : Contexts) {
// put content of while loop here
}
If that's what you thinking of then not sure that does what we want. For example for this:
namespace nm1 {
  namespace nm2 {
    class C {};
    class KernelName0 : public C {};
  } // namespace nm2
It will create a Contexts vector of size 1 containing only nm2. So that will not work.
Unless you are thinking of something else?

Reply:
Ah, it appears that doesn't do what I thought it would do.  

I was suggesting it because the entire business here with NSStr seems incorrect, we should just be outputting directly to a stream rather than using a string.  It likely requires making a stack of items, then for-eaching over them instead.  String prepends are just unbelievably expensive.